File: homework3/erlang/exercises.erl

-module (exercises).
-export([change/1]).
-export([sumOfCubesOfOdds/1]).
-export([powers/3]).

change(Amount) ->
    if Amount < 0 ->
       {false, "amount cannot be negative"};
    true ->
       Quarters = Amount div 25,
       AfterQuarters = Amount rem 25,
       Dimes = AfterQuarters div 10,
       AfterDimes = AfterQuarters rem 10,
       Nickels = AfterDimes div 5,
       Pennies = AfterDimes rem 5,
       {true, [Quarters, Dimes, Nickels, Pennies]}
    end.
  
sumOfCubesOfOdds(List)-> 
  lists:sum([X*X*X || X <- List, X rem 2 =/= 0]).
 
% My colleague Mansi helped me in writing this function(powers).
powers(Base, Limit, Pid)->
  nextPower(Base, Limit, Pid, 1).

nextPower(Base, Limit, Pid, Power) when Power =< Limit ->
  Pid ! Power,
  nextPower(Base, Limit, Pid, Power * Base);
nextPower(_, _,  _, _) ->
  ok.

File: homework3/go/exercises.go
package exercises

import (
	"errors"
	"fmt"
	"math"
)

type Cylinder struct {
	radius, height float64
}

func (c Cylinder) SurfaceArea() float64 {
	return 2.0*math.Pi*c.radius*c.height + 2.0*math.Pi*math.Pow(c.radius, 2)
}

func (c Cylinder) Volume() float64 {
	return math.Pi * math.Pow(c.radius, 2) * c.height
}

func Change(amount int) ([]int, error) {

	if amount < 0 {
		return nil, errors.New("amount cannot be negative")
	}

	var quarters = (amount / 25)
	amount = amount % 25

	var dimes = (amount / 10)
	amount = amount % 10

	var nickels = (amount / 5)

	var pennies = amount % 5

	var s = []int{quarters, dimes, nickels, pennies}
	return s, nil
}

func Powers(base int, limit int, ch chan<- int) {

	for power := 1; power <= limit; power *= base {
		ch <- power
	}

}

func main2() {
	ch := make(chan int, 100)
	go Powers(2, 70, ch)
	for p := range ch {
		fmt.Println(p)
	}

	close(ch)
}
